### PR TITLE
In custom facters extended key values (String, Array, Hash).

### PR DESCRIPTION
**Pull request by MemberIT**
Added support array and hash in key values of  custom facters.
code manifest:
```
  $title_facters = {
    roles => {
      ensure => present,
      value  => [ 'puppetserver', 'hiera', 'puppetdb', 'postgresql' ],
    },
  }
  create_resources('::puppet::fact', $title_facters)
```
result:
```
#  facter -v
3.3.0 (commit fc9282a2d426f7c8a742fd5b23c4cb06852eee20)
# facter roles
[
  "hiera",
  "postgresql",
  "puppetdb",
  "puppetserver"
]
```

### CODE REVIEW

**Comment by rendhalver** [comment]
Awesome idea.
Please adjusts the tests for this as well.
Thanks.

**Comment by MemberIT** [comment]
Done.

File: Gemfile [hunk 1]
```diff
@@ -19,6 +19,8 @@ group :test do
   gem 'metadata-json-lint',                               :require => false
   gem 'simplecov',                                        :require => false
   gem 'json',                   '1.8.3',                  :require => false
+  gem 'json_pure',              '1.8.3',                  :require => false
+  gem 'rubocop',                '0.41.2',                 :require => false
```
Comment (rendhalver) [inline Gemfile]: Thanks for that too. :)

File: manifests/fact.pp [hunk 2]
```diff
@@ -10,6 +10,10 @@
   include ::puppet::defaults
   $facterbasepath = $::puppet::defaults::facterbasepath
 
+  unless is_string($value) or is_array($value) or is_hash($value) {
+    fail("Param \$value of define Puppet::Fact must be Sring or Array or Hash!")
+  }
+
```
Comment (rendhalver) [inline manifests/fact.pp]: I think we should put this warning in puppet::facts as well.

**Comment by MemberIT** [comment]
When can i see my commits in master branch?

**Comment by rendhalver** [comment]
Sorry it's been a busy week.
Having a look now.

File: manifests/fact.pp [hunk 3]
```diff
@@ -10,6 +10,10 @@
   include ::puppet::defaults
   $facterbasepath = $::puppet::defaults::facterbasepath
 
+  unless is_string($value) or is_array($value) or is_hash($value) {
+    warning("Param \$value of define Puppet::Fact must be Sring or Array or Hash!")
+  }
+
```
Comment (rendhalver) [inline manifests/fact.pp]: Can you please add this into the puppet::facts class as well?
You have changed it's functionality as well so this warning is going to be needed there as well.

Comment (MemberIT) [inline manifests/fact.pp]: Ok. Tomorrow.

Comment (rendhalver) [inline manifests/fact.pp]: Thanks.

**Comment by MemberIT** [comment]
Hi, i closed this pull request:
feature is good, but code not good.
I will open later new pull request later.

**Pull request closed without merging**